src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs

```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
         /// <summary>
         /// FilterExpression corresponding to filter criteria
         /// </summary>
-        private FilterExpression filterExpression;
-        
+        private readonly FilterExpression filterExpression;
+
+        /// <remarks>
+        /// Exposed for testing purpose.
+        /// </remarks>
+        internal readonly FastFilter fastFilter;
+
+        private bool UseFilterExpression => this.filterExpression != null;
+
+        private bool UseFastFilter => this.fastFilter != null;
+
         /// <summary>
-        /// Initializes FilterExpressionWrapper with given filterString.  
+        /// Initializes FilterExpressionWrapper with given filterString and options.  
         /// </summary>
-        public FilterExpressionWrapper(string filterString)
+        public FilterExpressionWrapper(string filterString, FilterOptions options = null)
         {
             ValidateArg.NotNullOrEmpty(filterString, "filterString");
 
             this.FilterString = filterString;
+            this.FilterOptions = options;
+
             try
             {
-                this.filterExpression = FilterExpression.Parse(filterString);
+                // We prefer fast filter when it's available.
+                this.filterExpression = FilterExpression.Parse(filterString, out this.fastFilter);
+
+                if (UseFastFilter)
+                {
+                    this.filterExpression = null;
+
+                    // Property value regex is only supported for fast filter, 
+                    // so we ignore it if no fast filter is constructed.
+
+                    // TODO: surface an error message to suer.
+                    var regexString = options?.FilterRegEx;
+                    if (!string.IsNullOrEmpty(regexString))
+                    {
+                        this.fastFilter.PropertyValueRegex = new Regex(regexString, RegexOptions.Compiled);
+                    }
+                }
+
             }
             catch (FormatException ex)
             {
                 this.ParseError = ex.Message;
             }
+            catch (ArgumentException)
+            { 
+                // TODO: surface an error message to user.
+                // We don't match any tests in case of a regex parsing error.
+                this.fastFilter = null;
+            }
+        }
+
+        /// <summary>
+        /// Initializes FilterExpressionWrapper with given filterString.  
+        /// </summary>
+        public FilterExpressionWrapper(string filterString)
```

Comment:
After adding optional parameter `FilterOptions options = null` for FilterExpressionWrapper, this should be duplicated?

Reply:
This is for compatibility, since `FilterExpressionWrapper` is public.